<!DOCTYPE html>
<html lang="en"><head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="twitter:card" content="summary_large_image" /><!-- Begin Jekyll SEO tag v2.6.1 -->
<title>RNA-Seq analysis of SMOC2 using DESeq2 | Home</title>
<meta name="generator" content="Jekyll v4.1.1" />
<meta property="og:title" content="RNA-Seq analysis of SMOC2 using DESeq2" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="Complete RNA-Seq analysis workflow of SMOC2 over expression using DESeq2 and Tidyverse" />
<meta property="og:description" content="Complete RNA-Seq analysis workflow of SMOC2 over expression using DESeq2 and Tidyverse" />
<link rel="canonical" href="https://www.gurkamal.com//bioinformatics/r/2022/05/12/RNA-Seq-analysis-with-DESeq2.html" />
<meta property="og:url" content="https://www.gurkamal.com//bioinformatics/r/2022/05/12/RNA-Seq-analysis-with-DESeq2.html" />
<meta property="og:site_name" content="Home" />
<meta property="og:type" content="article" />
<meta property="article:published_time" content="2022-05-12T00:00:00-05:00" />
<script type="application/ld+json">
{"url":"https://www.gurkamal.com//bioinformatics/r/2022/05/12/RNA-Seq-analysis-with-DESeq2.html","@type":"BlogPosting","headline":"RNA-Seq analysis of SMOC2 using DESeq2","dateModified":"2022-05-12T00:00:00-05:00","datePublished":"2022-05-12T00:00:00-05:00","mainEntityOfPage":{"@type":"WebPage","@id":"https://www.gurkamal.com//bioinformatics/r/2022/05/12/RNA-Seq-analysis-with-DESeq2.html"},"description":"Complete RNA-Seq analysis workflow of SMOC2 over expression using DESeq2 and Tidyverse","@context":"https://schema.org"}</script>
<!-- End Jekyll SEO tag -->
<link rel="stylesheet" href="/assets/css/style.css"><link type="application/atom+xml" rel="alternate" href="https://www.gurkamal.com//feed.xml" title="Home" /><!-- the google_analytics_id gets auto inserted from the config file -->



<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-59WZ8TRRRR"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-59WZ8TRRRR');
</script>


<link rel="shortcut icon" type="image/x-icon" href="/images/favicon.ico"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/Primer/15.2.0/primer.css" integrity="sha512-xTz2ys4coGAOz8vuV1NcQBkgVmKhsSEtjbqyMJbBHRplFuvKIUo6xhLHpAyPt9mfR6twHJgn9OgVLuqOvjeBhg==" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.14.0/css/all.min.css" integrity="sha512-1PKOgIY59xJ8Co8+NE6FZ+LOAZKjy+KY8iq0G4B3CyeY6wYHN3yt9PW0XpSriVlkMXe40PTKnXrLnZ9+fkDaog==" crossorigin="anonymous" />
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.12.0/katex.min.css" integrity="sha512-h7nl+xz8wgDlNM4NqKEM4F1NkIRS17M9+uJwIGwuo8vGqIl4BhuCKdxjWEINm+xyrUjNCnK5dCrhM0sj+wTIXw==" crossorigin="anonymous" />
    <script src="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.12.0/katex.min.js" integrity="sha512-/CMIhXiDA3m2c9kzRyd97MTb3MC6OVnx4TElQ7fkkoRghwDf6gi41gaT1PwF270W6+J60uTmwgeRpNpJdRV6sg==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.12.0/contrib/auto-render.min.js" integrity="sha512-Do7uJAaHZm5OLrIv/yN4w0iG1dbu01kzdMNnFfu/mAqgUk6Nniv2JYHcwH+cNwjqgLcqcuBBk+JRvprLVI8azg==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha512-0doc9hKxR3PYwso42RD1p5ySZpzzuDiOwMrdCEh2WdJZCjcmFKc/wEnL+z8fBQrnHoiNWbo+3fiGkOYXBdQp4A==" crossorigin="anonymous"></script>
    <script>
    document.addEventListener("DOMContentLoaded", function() {
        renderMathInElement( document.body, {
        delimiters: [
            {left: "$$", right: "$$", display: true},
            {left: "[%", right: "%]", display: true},
            {left: "$", right: "$", display: false}
        ]}
        );
    });
    </script>


<script>
function wrap_img(fn) {
    if (document.attachEvent ? document.readyState === "complete" : document.readyState !== "loading") {
        var elements = document.querySelectorAll(".post img");
        Array.prototype.forEach.call(elements, function(el, i) {
            if (el.getAttribute("title") && (el.className != "emoji")) {
                const caption = document.createElement('figcaption');
                var node = document.createTextNode(el.getAttribute("title"));
                caption.appendChild(node);
                const wrapper = document.createElement('figure');
                wrapper.className = 'image';
                el.parentNode.insertBefore(wrapper, el);
                el.parentNode.removeChild(el);
                wrapper.appendChild(el);
                wrapper.appendChild(caption);
            }
        });
    } else { document.addEventListener('DOMContentLoaded', fn); }
}
window.onload = wrap_img;
</script>

<script>
    document.addEventListener("DOMContentLoaded", function(){
    // add link icon to anchor tags
    var elem = document.querySelectorAll(".anchor-link")
    elem.forEach(e => (e.innerHTML = '<i class="fas fa-link fa-xs"></i>'));
    });
</script>
</head>
<body><header class="site-header">

  <div class="wrapper"><a class="site-title" rel="author" href="/">Home</a><nav class="site-nav">
        <input type="checkbox" id="nav-trigger" class="nav-trigger" />
        <label for="nav-trigger">
          <span class="menu-icon">
            <svg viewBox="0 0 18 15" width="18px" height="15px">
              <path d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.032C17.335,0,18,0.665,18,1.484L18,1.484z M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.032C17.335,6.031,18,6.696,18,7.516L18,7.516z M18,13.516C18,14.335,17.335,15,16.516,15H1.484 C0.665,15,0,14.335,0,13.516l0,0c0-0.82,0.665-1.483,1.484-1.483h15.032C17.335,12.031,18,12.695,18,13.516L18,13.516z"/>
            </svg>
          </span>
        </label>

        <div class="trigger"><a class="page-link" href="/about/">About Me</a><a class="page-link" href="/search/">Search</a><a class="page-link" href="/categories/">Tags</a></div>
      </nav></div>
</header>
<main class="page-content" aria-label="Content">
      <div class="wrapper">
        <article class="post h-entry" itemscope itemtype="http://schema.org/BlogPosting">

  <header class="post-header">
    <h1 class="post-title p-name" itemprop="name headline">RNA-Seq analysis of SMOC2 using DESeq2</h1><p class="page-description">Complete RNA-Seq analysis workflow of SMOC2 over expression using DESeq2 and Tidyverse</p><p class="post-meta post-meta-title"><time class="dt-published" datetime="2022-05-12T00:00:00-05:00" itemprop="datePublished">
        May 12, 2022
      </time>
       • <span class="read-time" title="Estimated read time">
    
    
      28 min read
    
</span></p>

    
      <p class="category-tags"><i class="fas fa-tags category-tags-icon"></i></i> 
      
        <a class="category-tags-link" href="/categories/#Bioinformatics">Bioinformatics</a>
        &nbsp;
      
        <a class="category-tags-link" href="/categories/#R">R</a>
        
      
      </p>
    

    </header>

  <div class="post-content e-content" itemprop="articleBody">
    <ul class="section-nav">
<li class="toc-entry toc-h1"><a href="#introduction">Introduction</a>
<ul>
<li class="toc-entry toc-h3"><a href="#what-is-smoc2">What is smoc2?</a></li>
<li class="toc-entry toc-h3"><a href="#a-look-at-the-dataset">A look at the dataset</a></li>
<li class="toc-entry toc-h3"><a href="#tools-used">Tools used</a></li>
<li class="toc-entry toc-h3"><a href="#overview-of-workflow">Overview of workflow</a></li>
</ul>
</li>
<li class="toc-entry toc-h1"><a href="#importing-read-counts-associated-with-genes">Importing read counts associated with genes</a>
<ul>
<li class="toc-entry toc-h3"><a href="#importing-the-data">Importing the data</a></li>
<li class="toc-entry toc-h3"><a href="#putting-together-the-metadata">Putting together the metadata</a></li>
</ul>
</li>
<li class="toc-entry toc-h1"><a href="#normalization">Normalization</a>
<ul>
<li class="toc-entry toc-h3"><a href="#preparing-the-data-for-deseq2">Preparing the data for DESeq2</a></li>
<li class="toc-entry toc-h3"><a href="#creating-the-deseq2-dataset-dds">Creating the DESeq2 DataSet (DDS)</a></li>
<li class="toc-entry toc-h3"><a href="#estimating-size-factors">Estimating size factors</a></li>
<li class="toc-entry toc-h2"><a href="#extracting-normalized-counts">Extracting normalized counts</a>
<ul>
<li class="toc-entry toc-h3"><a href="#log-transform-the-normalized-counts">Log transform the normalized counts</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-entry toc-h1"><a href="#unsupervised-clustering-analysis">Unsupervised clustering analysis</a>
<ul>
<li class="toc-entry toc-h3"><a href="#hierarchical-clustering-with-a-correlation-heatmap">Hierarchical clustering with a correlation heatmap</a></li>
<li class="toc-entry toc-h3"><a href="#principal-component-analysis">Principal component analysis</a></li>
<li class="toc-entry toc-h3"><a href="#running-the-de-analysis">Running the DE analysis</a></li>
<li class="toc-entry toc-h3"><a href="#mean-and-variance">Mean and variance</a>
<ul>
<li class="toc-entry toc-h4"><a href="#relationship-between-mean-and-variance">Relationship between mean and variance</a></li>
</ul>
</li>
<li class="toc-entry toc-h3"><a href="#dispersion">Dispersion</a>
<ul>
<li class="toc-entry toc-h4"><a href="#plotting-dispersion-estimates">Plotting dispersion estimates</a></li>
</ul>
</li>
<li class="toc-entry toc-h3"><a href="#extracting-the-results-of-de-analysis">extracting the results of DE analysis</a></li>
<li class="toc-entry toc-h3"><a href="#ma-plot">MA plot</a></li>
<li class="toc-entry toc-h3"><a href="#lfc-shrinkage">LFC shrinkage</a></li>
<li class="toc-entry toc-h3"><a href="#create-a-results-dataframe">Create a results dataframe</a></li>
<li class="toc-entry toc-h3"><a href="#visualizing-results">Visualizing results</a></li>
</ul>
</li>
<li class="toc-entry toc-h1"><a href="#references">References</a></li>
</ul><h1 id="introduction">
<a class="anchor" href="#introduction" aria-hidden="true"><span class="octicon octicon-link"></span></a>Introduction</h1>

<h3 id="what-is-smoc2">
<a class="anchor" href="#what-is-smoc2" aria-hidden="true"><span class="octicon octicon-link"></span></a>What is smoc2?</h3>

<p>Secreted modular calcium-binding protein 2 (Smoc2) is an acidic extracellular matrix glycoprotein 
and plays an important role in bone mineralization, cell-matrix interactions, collagen binding, and 
bone remodeling[1]. SMOCs also influence cell growth factor signaling, cell migration and proliferation, 
angiogenesis. SMOC2 has been shown to have increased expression in kidney fibrosis, or renal fibrosis, 
which is characterized as an aberrant repair response to chronic tissue injury. This results in an excess 
of extracellular matrix in the space between tubules and capillaries within the kidney. However, it is 
unknown how Smoc2 functions in the induction and progression of fibrosis[2].</p>

<h3 id="a-look-at-the-dataset">
<a class="anchor" href="#a-look-at-the-dataset" aria-hidden="true"><span class="octicon octicon-link"></span></a>A look at the dataset</h3>

<p>This dataset cpmes from the research paper:</p>

<p><strong><em>Silencing SMOC2 ameliorates kidney fibrosis by inhibiting fibroblast to myofibroblast transformation.</em></strong></p>

<p>The experimental design:</p>

<ul>
  <li>
    <p>There are four sample groups being tested:</p>

    <ol>
      <li>
        <p>Normal control mice (wild type) with and without fibrosis</p>
      </li>
      <li>
        <p>Smoc2 over-expressing mice with and without fibrosis.</p>
      </li>
    </ol>
  </li>
  <li>
    <p>There are three biological replicates for all normal samples and four replicates for all fibrosis samples.</p>
  </li>
</ul>

<h3 id="tools-used">
<a class="anchor" href="#tools-used" aria-hidden="true"><span class="octicon octicon-link"></span></a>Tools used</h3>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1">#Load library for DESeq2</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">DESeq2</span><span class="p">)</span><span class="w">

</span><span class="c1">#Load library for RColorBrewer</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">RColorBrewer</span><span class="p">)</span><span class="w">

</span><span class="c1">#Load library for pheatmap</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">pheatmap</span><span class="p">)</span><span class="w">

</span><span class="c1">#Load library for tidyverse</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<h3 id="overview-of-workflow">
<a class="anchor" href="#overview-of-workflow" aria-hidden="true"><span class="octicon octicon-link"></span></a>Overview of workflow</h3>

<p>Differential expression analysis:</p>

<ol>
  <li>
    <p>Read in count matrix and create metadata</p>
  </li>
  <li>
    <p>Quality control:</p>
  </li>
</ol>

<ul>
  <li>
    <p>Normalize counts to account for differences in library depth.</p>
  </li>
  <li>
    <p>Perform principal component analysis</p>
  </li>
  <li>
    <p>Perform hierarchical clustering using heatmaps</p>
  </li>
  <li>
    <p>Identify potential sample outliers and major sources of variation in the data</p>
  </li>
</ul>

<ol>
  <li>
    <p>Differential expression analysis:</p>

    <ul>
      <li>
        <p>Run differential expression analysis</p>
      </li>
      <li>
        <p>Modeling counts</p>
      </li>
      <li>
        <p>Shrinking log2 fold change</p>
      </li>
      <li>
        <p>Testing for differential expression</p>
      </li>
    </ul>
  </li>
</ol>

<h1 id="importing-read-counts-associated-with-genes">
<a class="anchor" href="#importing-read-counts-associated-with-genes" aria-hidden="true"><span class="octicon octicon-link"></span></a>Importing read counts associated with genes</h1>

<h3 id="importing-the-data">
<a class="anchor" href="#importing-the-data" aria-hidden="true"><span class="octicon octicon-link"></span></a>Importing the data</h3>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read_csv</span><span class="p">(</span><span class="s2">"C:/Users/gurka/Downloads/fibrosis_smoc2_rawcounts/fibrosis_smoc2_rawcounts.csv"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>## New names:
## Rows: 47729 Columns: 8
## ── Column specification
## ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── Delimiter: "," chr
## (1): ...1 dbl (7): smoc2_fibrosis1, smoc2_fibrosis4, smoc2_normal1, smoc2_normal3, smoc2_fibrosis3, smoc2_normal4, smoc2_fibrosis2
## ℹ Use `spec()` to retrieve the full column specification for this data. ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
## • `` -&gt; `...1`
</code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">head</span><span class="p">(</span><span class="n">data</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>## # A tibble: 6 × 8
##   ...1               smoc2_fibrosis1 smoc2_fibrosis4 smoc2_normal1 smoc2_normal3 smoc2_fibrosis3 smoc2_normal4 smoc2_fibrosis2
##   &lt;chr&gt;                        &lt;dbl&gt;           &lt;dbl&gt;         &lt;dbl&gt;         &lt;dbl&gt;           &lt;dbl&gt;         &lt;dbl&gt;           &lt;dbl&gt;
## 1 ENSMUSG00000102693               0               0             0             0               0             0               0
## 2 ENSMUSG00000064842               0               0             0             0               0             0               0
## 3 ENSMUSG00000051951              72              30             0             3              36             1              51
## 4 ENSMUSG00000102851               0               0             0             0               0             0               0
## 5 ENSMUSG00000103377               0               0             1             0               0             0               0
## 6 ENSMUSG00000104017               0               0             0             0               0             0               0
</code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">subset</span><span class="w"> </span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="o">-</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">)])</span><span class="w">
</span></code></pre></div></div>

<h3 id="putting-together-the-metadata">
<a class="anchor" href="#putting-together-the-metadata" aria-hidden="true"><span class="octicon octicon-link"></span></a>Putting together the metadata</h3>

<p>Sample metadata is required in addition to the raw count data, data such as which of the samples correspond to each condition. To
create the metadata, a vector for each column is created and the vectors are combine into a data frame. The sample names are added as the row
names.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1">#create genotype vector</span><span class="w">
</span><span class="n">genotype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s1">'smoc2_oe'</span><span class="p">,</span><span class="s1">'smoc2_oe'</span><span class="p">,</span><span class="s1">'smoc2_oe'</span><span class="p">,</span><span class="s1">'smoc2_oe'</span><span class="p">,</span><span class="s1">'smoc2_oe'</span><span class="p">,</span><span class="s1">'smoc2_oe'</span><span class="p">,</span><span class="s1">'smoc2_oe'</span><span class="p">)</span><span class="w">

</span><span class="c1">#create condition vector</span><span class="w">
</span><span class="n">condition</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s1">'fibrosis'</span><span class="p">,</span><span class="s1">'fibrosis'</span><span class="p">,</span><span class="s1">'fibrosis'</span><span class="p">,</span><span class="s1">'fibrosis'</span><span class="p">,</span><span class="s1">'normal'</span><span class="p">,</span><span class="s1">'normal'</span><span class="p">,</span><span class="s1">'normal'</span><span class="p">)</span><span class="w">

</span><span class="c1">#create dataframe</span><span class="w">

</span><span class="n">smoc2_metadata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">genotype</span><span class="p">,</span><span class="w"> </span><span class="n">condition</span><span class="p">)</span><span class="w">

</span><span class="c1">#Assign the row names of the data frame</span><span class="w">
</span><span class="n">rownames</span><span class="p">(</span><span class="n">smoc2_metadata</span><span class="p">)</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s1">'smoc2_fibrosis1'</span><span class="p">,</span><span class="s1">'smoc2_fibrosis2'</span><span class="p">,</span><span class="s1">'smoc2_fibrosis3'</span><span class="p">,</span><span class="s1">'smoc2_fibrosis4'</span><span class="p">,</span><span class="w"> 
                            </span><span class="s1">'smoc2_normal1'</span><span class="p">,</span><span class="s1">'smoc2_normal3'</span><span class="p">,</span><span class="w"> </span><span class="s1">'smoc2_normal4'</span><span class="w"> </span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<h1 id="normalization">
<a class="anchor" href="#normalization" aria-hidden="true"><span class="octicon octicon-link"></span></a>Normalization</h1>

<p>Normalizing the raw counts to assess sample-level quality control is the first step in the workflow. The raw counts 
represent the number of reads aligning to each gene and should be proportional to the expression of the RNA in
the sample; however, there are factors other than RNA expression that can influence the number of reads aligning to each gene. 
The count data can be adjusted to remove the influence of these factors on the overall counts using normalization methods. 
The main factors often considered during normalization of count data are library depth, gene length, and RNA composition.</p>

<p>Differences in library size between samples can lead to many more reads being aligned to genes in one sample versus another sample.
Therefore, we need to adjust the counts assigned to each gene based on the size of the library prior to doing differential expression 
analysis.</p>

<p>Gene length normalization is another normalization factor often adjusted for. A longer gene generates a longer
transcript, which generates more fragments for sequencing. Therefore, a longer gene will often have more counts than a shorter
gene with the same level of expression. Since DE analysis compares expression levels of the same genes between conditions, we do 
not need to normalize for gene length. However, if you were to compare the expression levels of different genes, you would need 
to account for lengths of the genes.</p>

<p>Library composition effect is the third factor that needs to be taken into account for by adjusting library size, the composition 
of the library is also important. A few highly differentially expressed genes can skew many normalization methods that are not resistant 
to these outliers.</p>

<p>DESeq2 normalization uses a ‘median of ratios’ method of normalization. This method adjusts the raw counts for library size
and is resistant to large numbers of differentially expressed genes.</p>

<h3 id="preparing-the-data-for-deseq2">
<a class="anchor" href="#preparing-the-data-for-deseq2" aria-hidden="true"><span class="octicon octicon-link"></span></a>Preparing the data for DESeq2</h3>

<p>The data needs to be in a specific format to be accepted as an input to DESeq2. This requires the sample names in the 
metadata and counts datasets to be in the same order. Therefore, the row names of our metadata need to be in the same 
order as the column names of our counts data.</p>

<p>Currently the order is not the same, and this can be confirmed with the all() function to compare the row and coloumn order
which will return FALSE.The match() function can be used to determine the order needed. The first argumnent to the function 
is a vector of values in the order desired, and the second is a vector of values to be reordered.</p>

<p>Reordering can be done with the match() function by using the output of match() to reorder the rows of the metadata to be in
the same order as the columns in the count data. To do this the indices output by match() need to be extracted to a variable. 
Then, rearrange the metadata by using the square brackets and adding the index to the rows position. The samples should now be 
in the same order for both datasets, which can be confirmed by running the all() function once again</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1">#check if the column orders are the same with the all function</span><span class="w">
</span><span class="nf">all</span><span class="p">(</span><span class="n">rownames</span><span class="p">(</span><span class="n">smoc2_metadata</span><span class="p">)</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="n">colnames</span><span class="p">(</span><span class="n">data</span><span class="p">))</span><span class="w">

</span><span class="p">[</span><span class="m">1</span><span class="p">]</span><span class="w"> </span><span class="kc">FALSE</span><span class="w">
</span></code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1">#determine the matching order between the vectors using match()</span><span class="w">
</span><span class="n">match</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">data</span><span class="p">),</span><span class="w"> </span><span class="n">rownames</span><span class="p">(</span><span class="n">smoc2_metadata</span><span class="p">))</span><span class="w">

</span><span class="p">[</span><span class="m">1</span><span class="p">]</span><span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="m">4</span><span class="w"> </span><span class="m">5</span><span class="w"> </span><span class="m">6</span><span class="w"> </span><span class="m">3</span><span class="w"> </span><span class="m">7</span><span class="w"> </span><span class="m">2</span><span class="w">
</span></code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1">#use the match() function to reorder the column of the raw counts</span><span class="w">
</span><span class="n">reorder_idx</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">match</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">data</span><span class="p">),</span><span class="w"> </span><span class="n">rownames</span><span class="p">(</span><span class="n">smoc2_metadata</span><span class="p">))</span><span class="w">

</span><span class="c1">#reorder the metadata table</span><span class="w">
</span><span class="n">reordered_smoc2_meta</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">smoc2_metadata</span><span class="p">[</span><span class="n">reorder_idx</span><span class="p">,</span><span class="w"> </span><span class="p">]</span><span class="w">

</span><span class="c1">#view the new table</span><span class="w">
</span><span class="n">reordered_smoc2_meta</span><span class="w">

                     </span><span class="n">genotype</span><span class="w"> </span><span class="n">condition</span><span class="w">
     </span><span class="n">smoc2_fibrosis1</span><span class="w"> </span><span class="n">smoc2_oe</span><span class="w">  </span><span class="n">fibrosis</span><span class="w">
     </span><span class="n">smoc2_fibrosis4</span><span class="w"> </span><span class="n">smoc2_oe</span><span class="w">  </span><span class="n">fibrosis</span><span class="w">
     </span><span class="n">smoc2_normal1</span><span class="w">   </span><span class="n">smoc2_oe</span><span class="w">    </span><span class="n">normal</span><span class="w">
     </span><span class="n">smoc2_normal3</span><span class="w">   </span><span class="n">smoc2_oe</span><span class="w">    </span><span class="n">normal</span><span class="w">
     </span><span class="n">smoc2_fibrosis3</span><span class="w"> </span><span class="n">smoc2_oe</span><span class="w">  </span><span class="n">fibrosis</span><span class="w">
     </span><span class="n">smoc2_normal4</span><span class="w">   </span><span class="n">smoc2_oe</span><span class="w">    </span><span class="n">normal</span><span class="w">
     </span><span class="n">smoc2_fibrosis2</span><span class="w"> </span><span class="n">smoc2_oe</span><span class="w">  </span><span class="n">fibrosis</span><span class="w">
</span></code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1">#check to see if the column and row names are in the same order</span><span class="w">
</span><span class="nf">all</span><span class="p">(</span><span class="n">rownames</span><span class="p">(</span><span class="n">reordered_smoc2_meta</span><span class="p">)</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="n">colnames</span><span class="p">(</span><span class="n">data</span><span class="p">))</span><span class="w">

</span><span class="p">[</span><span class="m">1</span><span class="p">]</span><span class="w"> </span><span class="kc">TRUE</span><span class="w">
</span></code></pre></div></div>

<h3 id="creating-the-deseq2-dataset-dds">
<a class="anchor" href="#creating-the-deseq2-dataset-dds" aria-hidden="true"><span class="octicon octicon-link"></span></a>Creating the DESeq2 DataSet (DDS)</h3>

<p>Now a DESeq2 dataSet (DDS) needs to be created to store the raw counts and metadata to use for performing 
the differential expression analysis. The DESeq2 object is created with the <strong>DESeqDataSetFromMatrix()</strong> function.
This function takes as input the raw counts, associated metadata, and a design formula detailing which conditions 
in the metadata to use for differential expression analysis.</p>

<p>This function creates a DESeq2 object, of the class Ranged Summarized Experiment. This is a list-like object with slots 
available for the data it will generate throughout the analysis. Currently, it only has a few of the slots filled with the
count data, metadata, and design information.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Create a DESeq2 object</span><span class="w">
</span><span class="n">dds_smoc2</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">DESeqDataSetFromMatrix</span><span class="p">(</span><span class="n">countData</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">data</span><span class="p">,</span><span class="w">
                                   </span><span class="n">colData</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">reordered_smoc2_meta</span><span class="p">,</span><span class="w">
                                   </span><span class="n">design</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">condition</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<h3 id="estimating-size-factors">
<a class="anchor" href="#estimating-size-factors" aria-hidden="true"><span class="octicon octicon-link"></span></a>Estimating size factors</h3>

<p>The function estimateSizeFactors() can be used on the DESeq2 object to calculate the normalized counts and assign the output to a 
slot in the DESeq2 object, by re-assigning to DDS. DESeq2 will use these size factors to normalize the raw counts. The raw counts for each 
sample are divided by the associated sample-specific size factor for normalization. To view the size factors used for normalization, the 
sizeFactors() function can be used.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1">#estimating the size factors and feeding them back to the dds object by reassigning to the dds object</span><span class="w">
</span><span class="n">dds_smoc2</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">estimateSizeFactors</span><span class="p">(</span><span class="n">dds_smoc2</span><span class="p">)</span><span class="w">

</span><span class="c1">#viewing the size factors</span><span class="w">
</span><span class="n">sizeFactors</span><span class="p">(</span><span class="n">dds_smoc2</span><span class="p">)</span><span class="w">

     </span><span class="n">smoc2_fibrosis1</span><span class="w"> </span><span class="n">smoc2_fibrosis4</span><span class="w">   </span><span class="n">smoc2_normal1</span><span class="w">   </span><span class="n">smoc2_normal3</span><span class="w"> </span><span class="n">smoc2_fibrosis3</span><span class="w">   </span><span class="n">smoc2_normal4</span><span class="w"> </span><span class="n">smoc2_fibrosis2</span><span class="w"> 
           </span><span class="m">1.4319832</span><span class="w">       </span><span class="m">1.0826799</span><span class="w">       </span><span class="m">0.7106482</span><span class="w">       </span><span class="m">0.7989734</span><span class="w">       </span><span class="m">1.2480024</span><span class="w">       </span><span class="m">0.8482426</span><span class="w">       </span><span class="m">1.1189642</span><span class="w">
</span></code></pre></div></div>

<h2 id="extracting-normalized-counts">
<a class="anchor" href="#extracting-normalized-counts" aria-hidden="true"><span class="octicon octicon-link"></span></a>Extracting normalized counts</h2>

<p>Once the size factors have been calculated and added to the DESeq2 object, the normalized counts can be extracted from it. To extract 
the normalized counts from the DESeq2 object the <code class="language-plaintext highlighter-rouge">counts()</code> function can be used while specifying the normalized counts with the
<code class="language-plaintext highlighter-rouge">normalized = TRUE</code> argument. If the default was left as <code class="language-plaintext highlighter-rouge">normalized = FALSE</code>, then the raw counts would be extracted from the oject.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1">#once factors are calculated and reassigned then the normalized counts can be extracted using the counts function with normalized=TRUE argument</span><span class="w">
</span><span class="n">normalized_smoc2_counts</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">counts</span><span class="p">(</span><span class="n">dds_smoc2</span><span class="p">,</span><span class="w"> </span><span class="n">normalized</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">head</span><span class="p">(</span><span class="n">normalized_smoc2_counts</span><span class="p">)</span><span class="w">


          </span><span class="n">smoc2_fibrosis1</span><span class="w"> </span><span class="n">smoc2_fibrosis4</span><span class="w"> </span><span class="n">smoc2_normal1</span><span class="w"> </span><span class="n">smoc2_normal3</span><span class="w"> </span><span class="n">smoc2_fibrosis3</span><span class="w"> </span><span class="n">smoc2_normal4</span><span class="w"> </span><span class="n">smoc2_fibrosis2</span><span class="w">
     </span><span class="p">[</span><span class="m">1</span><span class="p">,]</span><span class="w">         </span><span class="m">0.00000</span><span class="w">         </span><span class="m">0.00000</span><span class="w">      </span><span class="m">0.000000</span><span class="w">      </span><span class="m">0.000000</span><span class="w">          </span><span class="m">0.0000</span><span class="w">      </span><span class="m">0.000000</span><span class="w">         </span><span class="m">0.00000</span><span class="w">
     </span><span class="p">[</span><span class="m">2</span><span class="p">,]</span><span class="w">         </span><span class="m">0.00000</span><span class="w">         </span><span class="m">0.00000</span><span class="w">      </span><span class="m">0.000000</span><span class="w">      </span><span class="m">0.000000</span><span class="w">          </span><span class="m">0.0000</span><span class="w">      </span><span class="m">0.000000</span><span class="w">         </span><span class="m">0.00000</span><span class="w">
     </span><span class="p">[</span><span class="m">3</span><span class="p">,]</span><span class="w">        </span><span class="m">50.27992</span><span class="w">        </span><span class="m">27.70902</span><span class="w">      </span><span class="m">0.000000</span><span class="w">      </span><span class="m">3.754818</span><span class="w">         </span><span class="m">28.8461</span><span class="w">      </span><span class="m">1.178908</span><span class="w">        </span><span class="m">45.57786</span><span class="w">
     </span><span class="p">[</span><span class="m">4</span><span class="p">,]</span><span class="w">         </span><span class="m">0.00000</span><span class="w">         </span><span class="m">0.00000</span><span class="w">      </span><span class="m">0.000000</span><span class="w">      </span><span class="m">0.000000</span><span class="w">          </span><span class="m">0.0000</span><span class="w">      </span><span class="m">0.000000</span><span class="w">         </span><span class="m">0.00000</span><span class="w">
     </span><span class="p">[</span><span class="m">5</span><span class="p">,]</span><span class="w">         </span><span class="m">0.00000</span><span class="w">         </span><span class="m">0.00000</span><span class="w">      </span><span class="m">1.407166</span><span class="w">      </span><span class="m">0.000000</span><span class="w">          </span><span class="m">0.0000</span><span class="w">      </span><span class="m">0.000000</span><span class="w">         </span><span class="m">0.00000</span><span class="w">
     </span><span class="p">[</span><span class="m">6</span><span class="p">,]</span><span class="w">         </span><span class="m">0.00000</span><span class="w">         </span><span class="m">0.00000</span><span class="w">      </span><span class="m">0.000000</span><span class="w">      </span><span class="m">0.000000</span><span class="w">          </span><span class="m">0.0000</span><span class="w">      </span><span class="m">0.000000</span><span class="w">         </span><span class="m">0.00000</span><span class="w">
</span></code></pre></div></div>

<h3 id="log-transform-the-normalized-counts">
<a class="anchor" href="#log-transform-the-normalized-counts" aria-hidden="true"><span class="octicon octicon-link"></span></a>Log transform the normalized counts</h3>

<p>Before visualizing the data the normalized counts should be log transformed to improve the visualization of the clustering.
For RNA-Seq data, DESeq2 uses a variance stabilizing transformation (VST), which is a logarithmic transformation that moderates the
variance across the mean. The normalized counts can be transformed by using the <code class="language-plaintext highlighter-rouge">DESeq2 vst()</code> function on the DESeq2 object. 
The <code class="language-plaintext highlighter-rouge">blind=TRUE</code> argument specifies that the transformation should be blind to the sample information given in the design formula; 
this argument should be specified when performing quality assessment.</p>

<h1 id="unsupervised-clustering-analysis">
<a class="anchor" href="#unsupervised-clustering-analysis" aria-hidden="true"><span class="octicon octicon-link"></span></a>Unsupervised clustering analysis</h1>

<p>With the counts normalized for library size, the counts between the different samples can now be compared. To assess the quality of 
experiment the samples are compared to each other with regards to gene expression. These comparisons are done visually though unsupervised
clustering analysis using hierarchical clustering heatmaps and PCA these are performed to get an idea of how similar the biological 
replicates are to each other and to identify outlier samples and major sources of variation present in the dataset.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1">#Unsupervised clustering analyses: log transformation</span><span class="w">

</span><span class="c1">#transform the normalized counts</span><span class="w">
</span><span class="n">vsd_smoc2</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">vst</span><span class="p">(</span><span class="n">dds_smoc2</span><span class="p">,</span><span class="w"> </span><span class="n">blind</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<h3 id="hierarchical-clustering-with-a-correlation-heatmap">
<a class="anchor" href="#hierarchical-clustering-with-a-correlation-heatmap" aria-hidden="true"><span class="octicon octicon-link"></span></a>Hierarchical clustering with a correlation heatmap</h3>

<p>Hierarchical clustering with heatmaps is used to assess the similarity in gene expression between the different samples in a dataset. This
technique is used to explore how similar replicates are to each other and whether the samples belonging to different sample groups
cluster separately. The heatmap is created by using the gene expression correlation values for all pairwise combinations of samples in the dataset, 
with the value 1 being perfect correlation. The hierarchical tree shows which samples are more similar to each other and the colors in the heatmap 
depict the correlation values. We expect the biological replicates to cluster together and sample conditions to cluster apart. Since the majority of 
genes should not be differentially expressed, samples should generally have high correlations with each other. Samples with correlation values below 
<strong>0.8</strong> may require further investigation to determine whether these samples are outliers or have contamination.</p>

<p>To create the heatmap the VST-transformed normalized counts will be extracted as a matrix from the vsd object using the <code class="language-plaintext highlighter-rouge">assay()</code> function. 
Then, the pairwise correlation values between each pair of samples will be computed using the <code class="language-plaintext highlighter-rouge">cor()</code> function</p>

<p>After generating the correlation values, the pheatmap package can be used to create the heatmap. The annotation argument selects which factors in the
metadata to include as annotation bars. Use the <code class="language-plaintext highlighter-rouge">select()</code> function from the <code class="language-plaintext highlighter-rouge">dplyr</code> package to select the condition column in the metadata.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1">#Hierarchical clustering with correlation heatmaps</span><span class="w">

</span><span class="c1">#extract the vst matrix (of transformed counts) from the object</span><span class="w">
</span><span class="n">vsd_mat_smoc2</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">assay</span><span class="p">(</span><span class="n">vsd_smoc2</span><span class="p">)</span><span class="w">

</span><span class="c1">#compute the pairwise correlation values between samples</span><span class="w">
</span><span class="n">vsd_cor_smoc2</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">cor</span><span class="p">(</span><span class="n">vsd_mat_smoc2</span><span class="p">)</span><span class="w">

</span><span class="c1">#view correlation statistics</span><span class="w">
</span><span class="n">vsd_cor_smoc2</span><span class="w">

                     </span><span class="n">smoc2_fibrosis1</span><span class="w"> </span><span class="n">smoc2_fibrosis4</span><span class="w"> </span><span class="n">smoc2_normal1</span><span class="w"> </span><span class="n">smoc2_normal3</span><span class="w"> </span><span class="n">smoc2_fibrosis3</span><span class="w"> </span><span class="n">smoc2_normal4</span><span class="w"> </span><span class="n">smoc2_fibrosis2</span><span class="w">
     </span><span class="n">smoc2_fibrosis1</span><span class="w">       </span><span class="m">1.0000000</span><span class="w">       </span><span class="m">0.9948457</span><span class="w">     </span><span class="m">0.9631782</span><span class="w">     </span><span class="m">0.9637385</span><span class="w">       </span><span class="m">0.9949621</span><span class="w">     </span><span class="m">0.9597660</span><span class="w">       </span><span class="m">0.9958622</span><span class="w">
     </span><span class="n">smoc2_fibrosis4</span><span class="w">       </span><span class="m">0.9948457</span><span class="w">       </span><span class="m">1.0000000</span><span class="w">     </span><span class="m">0.9668704</span><span class="w">     </span><span class="m">0.9671808</span><span class="w">       </span><span class="m">0.9951846</span><span class="w">     </span><span class="m">0.9635529</span><span class="w">       </span><span class="m">0.9946819</span><span class="w">
     </span><span class="n">smoc2_normal1</span><span class="w">         </span><span class="m">0.9631782</span><span class="w">       </span><span class="m">0.9668704</span><span class="w">     </span><span class="m">1.0000000</span><span class="w">     </span><span class="m">0.9935479</span><span class="w">       </span><span class="m">0.9678895</span><span class="w">     </span><span class="m">0.9940744</span><span class="w">       </span><span class="m">0.9651063</span><span class="w">
     </span><span class="n">smoc2_normal3</span><span class="w">         </span><span class="m">0.9637385</span><span class="w">       </span><span class="m">0.9671808</span><span class="w">     </span><span class="m">0.9935479</span><span class="w">     </span><span class="m">1.0000000</span><span class="w">       </span><span class="m">0.9691965</span><span class="w">     </span><span class="m">0.9942610</span><span class="w">       </span><span class="m">0.9656578</span><span class="w">
     </span><span class="n">smoc2_fibrosis3</span><span class="w">       </span><span class="m">0.9949621</span><span class="w">       </span><span class="m">0.9951846</span><span class="w">     </span><span class="m">0.9678895</span><span class="w">     </span><span class="m">0.9691965</span><span class="w">       </span><span class="m">1.0000000</span><span class="w">     </span><span class="m">0.9650676</span><span class="w">       </span><span class="m">0.9952111</span><span class="w">
     </span><span class="n">smoc2_normal4</span><span class="w">         </span><span class="m">0.9597660</span><span class="w">       </span><span class="m">0.9635529</span><span class="w">     </span><span class="m">0.9940744</span><span class="w">     </span><span class="m">0.9942610</span><span class="w">       </span><span class="m">0.9650676</span><span class="w">     </span><span class="m">1.0000000</span><span class="w">       </span><span class="m">0.9618977</span><span class="w">
     </span><span class="n">smoc2_fibrosis2</span><span class="w">       </span><span class="m">0.9958622</span><span class="w">       </span><span class="m">0.9946819</span><span class="w">     </span><span class="m">0.9651063</span><span class="w">     </span><span class="m">0.9656578</span><span class="w">       </span><span class="m">0.9952111</span><span class="w">     </span><span class="m">0.9618977</span><span class="w">       </span><span class="m">1.0000000</span><span class="w">
</span></code></pre></div></div>

<p>The output fromthe heatmap shows that the biological replicates cluster together and the conditions separate. This is encouraging since
differentially expressed genes between the conditions are likely to be driving this separation. Also, all correlation values are expectedly 
high without any outlier samples. If the replicates did not cluster as expected, the heatmap could be plotted with all of the metadata and 
see whether any other factor corresponds to the separation of the samples. If so, you would want to see if you get similar results with the 
PCA.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1">#plot the heatmap</span><span class="w">
</span><span class="n">pheatmap</span><span class="p">(</span><span class="n">vsd_cor_smoc2</span><span class="p">,</span><span class="w"> </span><span class="n">annotation</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">select</span><span class="p">(</span><span class="n">smoc2_metadata</span><span class="p">,</span><span class="w"> </span><span class="n">condition</span><span class="p">))</span><span class="w">
</span></code></pre></div></div>

<p><img src="images/heatmap.png" alt="plot the heatmap"></p>

<h3 id="principal-component-analysis">
<a class="anchor" href="#principal-component-analysis" aria-hidden="true"><span class="octicon octicon-link"></span></a>Principal component analysis</h3>

<p>PCA is a technique used to emphasize the variation present in a dataset. PCA finds the principal components of a dataset, with the first principal
component, or PC1, representing the greatest amount of variance in the data. Genes are given quantitative scores based on how much they influence 
the different PCs. A ‘per sample’ PC value is computed by taking the product of the influence and the normalized read count for each gene and summing 
across all genes.For PCA we generally plot these per sample PC values. Samples that cluster together have more similar gene expression profiles than 
samples that cluster apart, especially for the most variant genes.</p>

<p>This is a good method to explore the quality of the data as we hope to see replicates cluster together and conditions to separate on PC1. Sample 
outliers and major sources of variation can also be identified with this method.</p>

<p>Perform PCA using DESeq2’s <code class="language-plaintext highlighter-rouge">plotPCA()</code> function to plot the first two PCs. This function takes as input the transformed vsd object, and the
<code class="language-plaintext highlighter-rouge">intgroup</code> argument to specify what factor in the metadata to use to color the plot. The sample groups, normal and fibrosis, separate well on PC1. 
This means that the condition corresponds to PC1, which represents 88% of the variance in the data, while 4% is explained by PC2. This is great since 
it seems that a lot of the variation in gene expression in the dataset can likely be explained by the differences between sample groups. However, if 
the samples do not separate by PC1 or PC2, the effect of the condition could be small or there are sometimes other larger sources of variation present. 
You can color the PCA by other factors, such as age, sex, batch, etcetera, to identify other major sources of variation that could correspond to one 
of the large principal components.</p>

<p>Perform PCA to look how the samples cluster and whether the condition of interest corresponds with
the principal components explaining the most variation in the data.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1">#plot pca</span><span class="w">
</span><span class="n">plotPCA</span><span class="p">(</span><span class="n">vsd_smoc2</span><span class="p">,</span><span class="w"> </span><span class="n">intgroup</span><span class="o">=</span><span class="s2">"condition"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p><img src="images/deseq2pca.png" alt="plot pca"></p>

<h3 id="running-the-de-analysis">
<a class="anchor" href="#running-the-de-analysis" aria-hidden="true"><span class="octicon octicon-link"></span></a>Running the DE analysis</h3>

<p>Now that we have explored the quality of our samples, removed any outlier samples, and assessed the major sources of variation present, we can start the 
differential expression analysis with DESeq2. by performing the differential expression analysis with DESeq2, we are trying to identify which genes have 
significant differences in expression between the normal and fibrosis sample groups.</p>

<p>The differential expression analysis with DESeq2 consists of roughly three steps:</p>

<ol>
  <li>Fitting the raw counts for each gene to the DESeq2 negative binomial model and testing for differential expression.</li>
  <li>Shrinking the log2 fold changes.</li>
  <li>Extracting and visualizing the results.</li>
</ol>

<p>The design formula is an important part of modeling, telling DESeq2 the known major sources of variation to control for, or regress out, as well as, the condition 
of interest to use for differential expression testing. Also, the factor names in the design formula need to exactly match the column names in the metadata.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Run DESeq2 analysis </span><span class="w">
</span><span class="n">dds_smoc2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">DESeq</span><span class="p">(</span><span class="n">dds_smoc2</span><span class="p">)</span><span class="w">

     </span><span class="n">using</span><span class="w"> </span><span class="n">pre</span><span class="o">-</span><span class="n">existing</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="n">factors</span><span class="w">

     </span><span class="n">estimating</span><span class="w"> </span><span class="n">dispersions</span><span class="w">

     </span><span class="n">gene</span><span class="o">-</span><span class="n">wise</span><span class="w"> </span><span class="n">dispersion</span><span class="w"> </span><span class="n">estimates</span><span class="w">

     </span><span class="n">mean</span><span class="o">-</span><span class="n">dispersion</span><span class="w"> </span><span class="n">relationship</span><span class="w">

     </span><span class="n">final</span><span class="w"> </span><span class="n">dispersion</span><span class="w"> </span><span class="n">estimates</span><span class="w">

     </span><span class="n">fitting</span><span class="w"> </span><span class="n">model</span><span class="w"> </span><span class="n">and</span><span class="w"> </span><span class="n">testing</span><span class="w">
</span></code></pre></div></div>

<h3 id="mean-and-variance">
<a class="anchor" href="#mean-and-variance" aria-hidden="true"><span class="octicon octicon-link"></span></a>Mean and variance</h3>

<p>explore the fit of the smoc2 data to the negative binomial model by plotting the dispersion estimates using the <code class="language-plaintext highlighter-rouge">plotDispEsts()</code> function. The dispersion 
estimates are used to model the raw counts; if the dispersions don’t follow the assumptions made by DESeq2, then the variation in the data could be poorly estimated 
and the DE results could be less accurate.</p>

<p>The assumptions DESeq2 makes are that the dispersions should generally decrease with increasing mean and that they should more or less follow
the fitted line. The goal of the differential expression analysis is to determine whether a gene’s mean expression between sample groups is
different given the variation within groups. This is determined by testing the probability of the log2 fold changes between groups being significantly different from zero. 
The log2 fold changes are found by the log of the one sample group mean, therefore, to model the counts requires information about the mean and variation in the data. To
explore the variation in the data, the variance will be observed in gene expression relative to the mean. Variance is the square of the standard deviation, representing 
how far away the expression of the individual samples are from the means.</p>

<p>For RNA-Seq data, the variance is generally expected to increase with the gene’s mean expression. To observe this relationship, calculate the means and variances for
every gene of the normal samples using the <code class="language-plaintext highlighter-rouge">apply()</code> function.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Calculating mean for each gene (each row)</span><span class="w">
</span><span class="n">mean_counts</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">[,</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">3</span><span class="p">],</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">mean</span><span class="p">)</span><span class="w">
</span><span class="n">head</span><span class="p">(</span><span class="n">mean_counts</span><span class="p">)</span><span class="w">

</span><span class="p">[</span><span class="m">1</span><span class="p">]</span><span class="w">  </span><span class="m">0.0000000</span><span class="w">  </span><span class="m">0.0000000</span><span class="w"> </span><span class="m">34.0000000</span><span class="w">  </span><span class="m">0.0000000</span><span class="w">  </span><span class="m">0.3333333</span><span class="w">  </span><span class="m">0.0000000</span><span class="w">
</span></code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Calculating variance for each gene( each row)</span><span class="w">
</span><span class="n">variance_counts</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">apply</span><span class="p">(</span><span class="n">data</span><span class="p">[,</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">3</span><span class="p">],</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">var</span><span class="p">)</span><span class="w">
</span><span class="n">head</span><span class="p">(</span><span class="n">variance_counts</span><span class="p">)</span><span class="w">

</span><span class="p">[</span><span class="m">1</span><span class="p">]</span><span class="w">    </span><span class="m">0.0000000</span><span class="w">    </span><span class="m">0.0000000</span><span class="w"> </span><span class="m">1308.0000000</span><span class="w">    </span><span class="m">0.0000000</span><span class="w">    </span><span class="m">0.3333333</span><span class="w">    </span><span class="m">0.0000000</span><span class="w">
</span></code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1">#Plotting relationship between mean and variance:</span><span class="w">

</span><span class="c1"># Creating data frame with mean and vairance for every gene</span><span class="w">
</span><span class="n">df</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">mean_counts</span><span class="p">,</span><span class="w"> </span><span class="n">variance_counts</span><span class="p">)</span><span class="w">
</span><span class="n">head</span><span class="p">(</span><span class="n">df</span><span class="p">)</span><span class="w">

       </span><span class="n">mean_counts</span><span class="w"> </span><span class="n">variance_counts</span><span class="w">
     </span><span class="m">1</span><span class="w">   </span><span class="m">0.0000000</span><span class="w">       </span><span class="m">0.0000000</span><span class="w">
     </span><span class="m">2</span><span class="w">   </span><span class="m">0.0000000</span><span class="w">       </span><span class="m">0.0000000</span><span class="w">
     </span><span class="m">3</span><span class="w">  </span><span class="m">34.0000000</span><span class="w">    </span><span class="m">1308.0000000</span><span class="w">
     </span><span class="m">4</span><span class="w">   </span><span class="m">0.0000000</span><span class="w">       </span><span class="m">0.0000000</span><span class="w">
     </span><span class="m">5</span><span class="w">   </span><span class="m">0.3333333</span><span class="w">       </span><span class="m">0.3333333</span><span class="w">
     </span><span class="m">6</span><span class="w">   </span><span class="m">0.0000000</span><span class="w">       </span><span class="m">0.0000000</span><span class="w">
</span></code></pre></div></div>

<p>Create a data frame for plotting with <code class="language-plaintext highlighter-rouge">ggplot2</code>. Plot the mean and variance values for each gene using log10 scales. Each black dot represents a gene.</p>

<h4 id="relationship-between-mean-and-variance">
<a class="anchor" href="#relationship-between-mean-and-variance" aria-hidden="true"><span class="octicon octicon-link"></span></a>Relationship between mean and variance</h4>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">df</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">mean_counts</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="o">=</span><span class="n">variance_counts</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_y_log10</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_x_log10</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">xlab</span><span class="p">(</span><span class="s2">"Mean counts per gene"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">ylab</span><span class="p">(</span><span class="s2">"Variance per gene"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p><img src="images/dispersion.png" alt=""></p>

<p>The variance in gene expression increases with the mean. This is expected for RNA-Seq data. Also, note how the range in values for variance is greater for 
lower mean counts than higher mean counts. This is also expected for RNA-Seq count data.</p>

<h3 id="dispersion">
<a class="anchor" href="#dispersion" aria-hidden="true"><span class="octicon octicon-link"></span></a>Dispersion</h3>

<p>A measure of the variance for a given mean is described by a metric called dispersion in the DESeq2 model. The DESeq2 model uses dispersion to assess the variability in 
expression when modeling the counts.</p>

<p>The DESeq2 model calculates dispersion as being indirectly related to the mean and directly related to the variance of the  data using the formula displayed. 
So, an increase in variance will increase dispersion, while an increase in mean will decrease dispersion. For any two genes with the same mean expression, 
the only difference in dispersion will be based on differences in variance. To check the fit of the data to the DESeq2 model, it can be useful to look at 
the dispersion estimates.</p>

<h4 id="plotting-dispersion-estimates">
<a class="anchor" href="#plotting-dispersion-estimates" aria-hidden="true"><span class="octicon octicon-link"></span></a>Plotting dispersion estimates</h4>

<p>To plot the dispersions relative to the means for each gene, use the <code class="language-plaintext highlighter-rouge">plotDispEsts()</code> function on the DESeq2 object. Each black dot is a gene with associated 
mean and dispersion values. Expect dispersion values to decrease with increasing mean, which is what is seen. With only a few replicates for RNA-Seq experiments, 
gene-wise estimates of dispersion are often inaccurate, so DESeq2 uses information across all genes to determine the most likely estimates of dispersion for a 
given mean expression value, shown with the red line in the figure. Genes with inaccurately small estimates of variation could yield many false positives, or 
genes that are identified as DE, when they are really not. Therefore, the original gene-wise dispersion estimates, shown as the black dots in the figure, are 
shrunken towards the curve to yield more accurate estimates of dispersion, shown as blue dots. The more accurate, shrunken dispersion estimates are used to model the
counts for determining the differentially-expressed genes. Extremely high dispersion values, shown surrounded by blue circles, are not shrunken, due to the likelihood 
that the gene may have higher variability than others for biological or technical reasons and reducing the variation could result in false positives. The strength
of the shrinkage is dependent on the distance from the curve and sample size. Larger numbers of replicates can estimate the mean and variation more accurately, 
so yield less shrinkage.</p>

<p>Worrisome plots would include a cloud of data that doesn’t follow the curve or dispersions that don’t decrease with increasing mean. These problems can often be explained 
by sample outliers or contamination.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1">#plot dispersion estimates</span><span class="w">
</span><span class="n">plotDispEsts</span><span class="p">(</span><span class="n">dds_smoc2</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>
<p>The dispersion estimates are used to model the raw counts; if the dispersions don’t follow the assumptions made by DESeq2, then the
variation in the data could be poorly estimated and the DE results could be less accurate. The assumptions DESeq2 makes are that the dispersions should generally decrease with increasing mean and that they should more or less follow
the fitted line.</p>

<p><img src="images/shrink.png" alt=""></p>

<h3 id="extracting-the-results-of-de-analysis">
<a class="anchor" href="#extracting-the-results-of-de-analysis" aria-hidden="true"><span class="octicon octicon-link"></span></a>extracting the results of DE analysis</h3>

<p>DESeq2 will perform the Wald test for pairwise comparisons to test for differences in expression between
two sample groups for the condition of interest, in this case, condition. The sample groups for condition 
are fibrosis and normal. The results of the testing can be extracted using the <code class="language-plaintext highlighter-rouge">results()</code> function and 
specifying a significance level, or alpha value, using the <code class="language-plaintext highlighter-rouge">alpha</code> argument. You can choose the alpha 
based on how stringent you want to be with your analysis. Lower alpha values indicate less probability of 
identifying a gene as DE when it is actually not. We will use a standard alpha of 0-point-05. The top of 
the output shows <strong>“log2 foldchange condition normal versus fibrosis”</strong>, indicating that fibrosis is the 
base level of comparison. This means that all log2 fold changes represent the normal group relative to the 
fibrosis group.</p>

<p>Contrasts specify the sample groups to compare and can be given directly to the <code class="language-plaintext highlighter-rouge">results()</code> function 
using the <code class="language-plaintext highlighter-rouge">contrast</code> argument. Within the combine function, we need to specify the condition of interest, 
level to compare, and base level. To specify the normal sample group as the base level for condition, we
could create the contrast defining normal as the base level and fibrosis as the level to compare. Note that 
the condition of interest and sample groups for the condition need to match the names in the metadata.</p>

<p>Now the results give log2 fold changes of the fibrosis group relative to normal.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># The syntax for DESeq2 contrasts is</span><span class="w">

</span><span class="c1"># results(dds,</span><span class="w">
</span><span class="c1">#         contrast = c("condition_factor", "level_to_compare", "base_level"),</span><span class="w">
</span><span class="c1">#         alpha = 0.05)</span><span class="w">

</span><span class="n">smoc2_results</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">results</span><span class="p">(</span><span class="n">dds_smoc2</span><span class="p">,</span><span class="w">
                        </span><span class="n">contrast</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"condition"</span><span class="p">,</span><span class="w"> </span><span class="s2">"fibrosis"</span><span class="p">,</span><span class="w"> </span><span class="s2">"normal"</span><span class="p">),</span><span class="w">
                        </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.05</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">head</span><span class="p">(</span><span class="n">smoc2_results</span><span class="p">)</span><span class="w">

     </span><span class="n">log2</span><span class="w"> </span><span class="n">fold</span><span class="w"> </span><span class="n">change</span><span class="w"> </span><span class="p">(</span><span class="n">MLE</span><span class="p">)</span><span class="o">:</span><span class="w"> </span><span class="n">condition</span><span class="w"> </span><span class="n">fibrosis</span><span class="w"> </span><span class="n">vs</span><span class="w"> </span><span class="n">normal</span><span class="w"> 
     </span><span class="n">Wald</span><span class="w"> </span><span class="n">test</span><span class="w"> </span><span class="n">p</span><span class="o">-</span><span class="n">value</span><span class="o">:</span><span class="w"> </span><span class="n">condition</span><span class="w"> </span><span class="n">fibrosis</span><span class="w"> </span><span class="n">vs</span><span class="w"> </span><span class="n">normal</span><span class="w"> 
     </span><span class="n">DataFrame</span><span class="w"> </span><span class="n">with</span><span class="w"> </span><span class="m">6</span><span class="w"> </span><span class="n">rows</span><span class="w"> </span><span class="n">and</span><span class="w"> </span><span class="m">6</span><span class="w"> </span><span class="n">columns</span><span class="w">
        </span><span class="n">baseMean</span><span class="w"> </span><span class="n">log2FoldChange</span><span class="w">     </span><span class="n">lfcSE</span><span class="w">      </span><span class="n">stat</span><span class="w">      </span><span class="n">pvalue</span><span class="w">        </span><span class="n">padj</span><span class="w">
       </span><span class="o">&lt;</span><span class="n">numeric</span><span class="o">&gt;</span><span class="w">      </span><span class="o">&lt;</span><span class="n">numeric</span><span class="o">&gt;</span><span class="w"> </span><span class="o">&lt;</span><span class="n">numeric</span><span class="o">&gt;</span><span class="w"> </span><span class="o">&lt;</span><span class="n">numeric</span><span class="o">&gt;</span><span class="w">   </span><span class="o">&lt;</span><span class="n">numeric</span><span class="o">&gt;</span><span class="w">   </span><span class="o">&lt;</span><span class="n">numeric</span><span class="o">&gt;</span><span class="w">
     </span><span class="m">1</span><span class="w">  </span><span class="m">0.000000</span><span class="w">             </span><span class="kc">NA</span><span class="w">        </span><span class="kc">NA</span><span class="w">        </span><span class="kc">NA</span><span class="w">          </span><span class="kc">NA</span><span class="w">          </span><span class="kc">NA</span><span class="w">
     </span><span class="m">2</span><span class="w">  </span><span class="m">0.000000</span><span class="w">             </span><span class="kc">NA</span><span class="w">        </span><span class="kc">NA</span><span class="w">        </span><span class="kc">NA</span><span class="w">          </span><span class="kc">NA</span><span class="w">          </span><span class="kc">NA</span><span class="w">
     </span><span class="m">3</span><span class="w"> </span><span class="m">22.478090</span><span class="w">        </span><span class="m">4.49814</span><span class="w">  </span><span class="m">0.829291</span><span class="w">  </span><span class="m">5.424085</span><span class="w"> </span><span class="m">5.82520e-08</span><span class="w"> </span><span class="m">2.54201e-07</span><span class="w">
     </span><span class="m">4</span><span class="w">  </span><span class="m">0.000000</span><span class="w">             </span><span class="kc">NA</span><span class="w">        </span><span class="kc">NA</span><span class="w">        </span><span class="kc">NA</span><span class="w">          </span><span class="kc">NA</span><span class="w">          </span><span class="kc">NA</span><span class="w">
     </span><span class="m">5</span><span class="w">  </span><span class="m">0.201024</span><span class="w">       </span><span class="m">-1.59170</span><span class="w">  </span><span class="m">3.816946</span><span class="w"> </span><span class="m">-0.417009</span><span class="w"> </span><span class="m">6.76672e-01</span><span class="w">          </span><span class="kc">NA</span><span class="w">
     </span><span class="m">6</span><span class="w">  </span><span class="m">0.000000</span><span class="w">             </span><span class="kc">NA</span><span class="w">        </span><span class="kc">NA</span><span class="w">        </span><span class="kc">NA</span><span class="w">          </span><span class="kc">NA</span><span class="w">          </span><span class="kc">NA</span><span class="w">
</span></code></pre></div></div>

<h3 id="ma-plot">
<a class="anchor" href="#ma-plot" aria-hidden="true"><span class="octicon octicon-link"></span></a>MA plot</h3>

<p>The MA plot shows the  mean of the normalized counts versus the log2 fold changes for all genes tested. 
The DESeq2 function plotMA() can be used to create the plot and the genes that are significantly DE are colored red. 
Note the large log2 foldchanges, particularly for genes with lower mean count values. 
These fold changes are unlikely to be as accurate for genes that have little information associated with them, 
such as genes with low numbers of counts or high dispersion values.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">plotMA</span><span class="p">(</span><span class="n">smoc2_results</span><span class="p">,</span><span class="w"> </span><span class="n">ylim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">-8</span><span class="p">,</span><span class="m">8</span><span class="p">))</span><span class="w">
</span></code></pre></div></div>

<p><img src="normalized.png" alt=""></p>

<h3 id="lfc-shrinkage">
<a class="anchor" href="#lfc-shrinkage" aria-hidden="true"><span class="octicon octicon-link"></span></a>LFC shrinkage</h3>

<p>To improve the estimated fold changes we can use log2 foldchange shrinkage. For genes with low amounts of information available, shrinkage uses information from all genes to generate more likely, lower, log2 fold change estimates, similar to what we did with dispersions. DESeq2 has the <code class="language-plaintext highlighter-rouge">lfcShrink()</code> function to generate the shrunken log2 foldchanges. We need to specify the DESeq2 object, the contrast, and our results object.</p>

<p>These shrunken log2 foldchanges should be more accurate; however, shrinking the log2 foldchanges will not affect the number of differentially expressed genes returned, only the log2 fold change values. Now that we have accurate fold changes, we can extract the significant DE genes and perform further visualizations of results.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">smoc2_results</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lfcShrink</span><span class="p">(</span><span class="n">dds_smoc2</span><span class="p">,</span><span class="w">
                          </span><span class="n">contrast</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"condition"</span><span class="p">,</span><span class="w"> </span><span class="s2">"fibrosis"</span><span class="p">,</span><span class="w"> </span><span class="s2">"normal"</span><span class="p">),</span><span class="w">
                          </span><span class="n">type</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'ashr'</span><span class="p">,</span><span class="w">
                          </span><span class="n">res</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">smoc2_results</span><span class="p">)</span><span class="w">

                          
     </span><span class="n">using</span><span class="w"> </span><span class="s1">'ashr'</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">LFC</span><span class="w"> </span><span class="n">shrinkage.</span><span class="w"> </span><span class="n">If</span><span class="w"> </span><span class="n">used</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="n">published</span><span class="w"> </span><span class="n">research</span><span class="p">,</span><span class="w"> </span><span class="n">please</span><span class="w"> </span><span class="n">cite</span><span class="o">:</span><span class="w">
         </span><span class="n">Stephens</span><span class="p">,</span><span class="w"> </span><span class="n">M.</span><span class="w"> </span><span class="p">(</span><span class="m">2016</span><span class="p">)</span><span class="w"> </span><span class="n">False</span><span class="w"> </span><span class="n">discovery</span><span class="w"> </span><span class="n">rates</span><span class="o">:</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">new</span><span class="w"> </span><span class="n">deal.</span><span class="w"> </span><span class="n">Biostatistics</span><span class="p">,</span><span class="w"> </span><span class="m">18</span><span class="o">:</span><span class="m">2</span><span class="n">.</span><span class="w">
         </span><span class="n">https</span><span class="o">://</span><span class="n">doi.org</span><span class="o">/</span><span class="m">10.1093</span><span class="o">/</span><span class="n">biostatistics</span><span class="o">/</span><span class="n">kxw041</span><span class="w">
</span></code></pre></div></div>

<h3 id="create-a-results-dataframe">
<a class="anchor" href="#create-a-results-dataframe" aria-hidden="true"><span class="octicon octicon-link"></span></a>Create a results dataframe</h3>

<p>Now that we have extracted our results, we can get a nice overview of the number of differentially expressed genes there are for our designated alpha level using the summary() function. It will output the numbers/percentages of up- and down-regulated genes, as well as, give information about independent filtering and outliers removed.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">summary</span><span class="p">(</span><span class="n">smoc2_results</span><span class="p">)</span><span class="w">

     </span><span class="n">out</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="m">29556</span><span class="w"> </span><span class="n">with</span><span class="w"> </span><span class="n">nonzero</span><span class="w"> </span><span class="n">total</span><span class="w"> </span><span class="n">read</span><span class="w"> </span><span class="n">count</span><span class="w">
     </span><span class="n">adjusted</span><span class="w"> </span><span class="n">p</span><span class="o">-</span><span class="n">value</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="m">0.05</span><span class="w">
     </span><span class="n">LFC</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="p">(</span><span class="n">up</span><span class="p">)</span><span class="w">       </span><span class="o">:</span><span class="w"> </span><span class="m">5776</span><span class="p">,</span><span class="w"> </span><span class="m">20</span><span class="o">%
     LFC &lt; 0 (down)     : 5332, 18%</span><span class="w">
     </span><span class="n">outliers</span><span class="w"> </span><span class="p">[</span><span class="m">1</span><span class="p">]</span><span class="w">       </span><span class="o">:</span><span class="w"> </span><span class="m">15</span><span class="p">,</span><span class="w"> </span><span class="m">0.051</span><span class="o">%
     low counts [2]     : 7207, 24%</span><span class="w">
     </span><span class="p">(</span><span class="n">mean</span><span class="w"> </span><span class="n">count</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="m">1</span><span class="p">)</span><span class="w">
     </span><span class="p">[</span><span class="m">1</span><span class="p">]</span><span class="w"> </span><span class="n">see</span><span class="w"> </span><span class="s1">'cooksCutoff'</span><span class="w"> </span><span class="n">argument</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="o">?</span><span class="n">results</span><span class="w">
     </span><span class="p">[</span><span class="m">2</span><span class="p">]</span><span class="w"> </span><span class="n">see</span><span class="w"> </span><span class="s1">'independentFiltering'</span><span class="w"> </span><span class="n">argument</span><span class="w"> </span><span class="n">of</span><span class="w"> </span><span class="o">?</span><span class="n">results</span><span class="w">
</span></code></pre></div></div>
<p>DESeq2’s <code class="language-plaintext highlighter-rouge">summary()</code> function provides the number of differentially expressed genes for the alpha level and information about the number of genes filtered. Our results give
over 10,000 genes as DE, which is the sum of the DE genes with log2 fold changes less than 0 and those with fold changes greater than 0. This is a lot of genes to sift through. If we wanted to return the genes most likely to be biologically relevant, we could also include a log2 fold change threshold. Oftentimes, a log2 fold change threshold isn’t preferred. However, it can be helpful when dealing with such large numbers of DE genes.</p>

<p>To test for significant genes using both an alpha value threshold and a log2 foldchange threshold different from 0, we need to re-run the results function.
Let’s use a small 1-point-25-foldchange threshold, which equals 0-point-32 on the log2 scale, by adding the lfcThreshold argument to our results() function. 
While using any log2 fold change cut-off increases the risk of losing biologically relevant genes, by using a very small log2 foldchange threshold, we are 
hoping to reduce the risk that the genes more biologically meaningful.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">smoc2_results</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">results</span><span class="p">(</span><span class="n">dds_smoc2</span><span class="p">,</span><span class="w"> </span><span class="n">contrast</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"condition"</span><span class="p">,</span><span class="w"> </span><span class="s2">"fibrosis"</span><span class="p">,</span><span class="w"> </span><span class="s2">"normal"</span><span class="p">),</span><span class="w"> </span><span class="n">lfcThreshold</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.32</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.05</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>Now we implemented a log2 fold change threshold of 1.25 fold (log2 0.32) when testing for significant genes. Now we can use these results to subset only the significant genes with p-adjusted values less than 0.05.</p>

<p>Now that we have the results, we need to re-shrink the foldchanges, then run the summary() function again. Now, we have returned just over 6,000 DE genes.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Save results as a data frame</span><span class="w">
</span><span class="n">smoc2_res_all</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">smoc2_results</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>Now let’s look at the values in the results table and identify the differentially expressed genes. To determine significant DE genes, we will be using the p-values adjusted for multiple test correction in the last column. The reason for this is that for every gene tested with an alpha of 0-point-05, there is a 5% chance that the gene is called as DE when it is not, yielding false positives. If we were to test the roughly 47,000 genes in the raw counts file, we would have about 5% or over 2,000 genes as false positives. It would be difficult to identify the true positives, or genes that are called DE when they truly are, from the false. Therefore, multiple test correction is performed by DESeq2 using the
Benjamini-Hochberg, or BH-method, to adjust p-values for multiple testing and control the proportion of false positives relative to true. Using the BH-method and an alpha value of 0-point-05, if we had 1,000 genes identified as DE, we would expect 5% of the DE genes to be false positives, or 50 genes. To reduce the number of genes tested, DESeq2 automatically filters out genes unlikely to be truly differentially expressed prior to testing, such as genes with zero counts across all samples, genes with low mean values across all samples, and genes with extreme count outliers.</p>

<p>To extract the significant DE genes, we’ll subset the results table, using the subset() function, for genes with p-adjusted values less than the alpha value of 0-point-05. We should see all p-adjusted values are less than 0-point-05 and log2 foldchanges are greater than the absolute value of 0-point-32. We will use the arrange() function to order the genes by p-adjusted values to generate the final table of significant results. We can explore this table for interesting or expected genes with a high probability of being related to kidney fibrosis.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Subset the results to only return the significant genes with p-adjusted values less than 0.05</span><span class="w">
</span><span class="n">smoc2_res_sig</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">subset</span><span class="p">(</span><span class="n">smoc2_res_all</span><span class="p">,</span><span class="w"> </span><span class="n">padj</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="m">0.05</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">head</span><span class="p">(</span><span class="n">smoc2_res_sig</span><span class="p">)</span><span class="w">

          </span><span class="n">baseMean</span><span class="w"> </span><span class="n">log2FoldChange</span><span class="w">      </span><span class="n">lfcSE</span><span class="w">       </span><span class="n">stat</span><span class="w">       </span><span class="n">pvalue</span><span class="w">         </span><span class="n">padj</span><span class="w">
     </span><span class="m">3</span><span class="w">    </span><span class="m">22.47809</span><span class="w">      </span><span class="m">4.4981432</span><span class="w"> </span><span class="m">0.82929064</span><span class="w">   </span><span class="m">5.038213</span><span class="w"> </span><span class="m">4.698974e-07</span><span class="w"> </span><span class="m">2.829520e-06</span><span class="w">
     </span><span class="m">17</span><span class="w">   </span><span class="m">12.06950</span><span class="w">     </span><span class="m">-2.3959881</span><span class="w"> </span><span class="m">0.60066414</span><span class="w">  </span><span class="m">-3.456155</span><span class="w"> </span><span class="m">5.479409e-04</span><span class="w"> </span><span class="m">2.278479e-03</span><span class="w">
     </span><span class="m">33</span><span class="w"> </span><span class="m">1380.35712</span><span class="w">     </span><span class="m">-0.8942696</span><span class="w"> </span><span class="m">0.09748260</span><span class="w">  </span><span class="m">-5.890996</span><span class="w"> </span><span class="m">3.838750e-09</span><span class="w"> </span><span class="m">2.815588e-08</span><span class="w">
     </span><span class="m">35</span><span class="w"> </span><span class="m">2522.97515</span><span class="w">     </span><span class="m">-1.9163511</span><span class="w"> </span><span class="m">0.14944995</span><span class="w"> </span><span class="m">-10.681510</span><span class="w"> </span><span class="m">1.242343e-26</span><span class="w"> </span><span class="m">2.749899e-25</span><span class="w">
     </span><span class="m">40</span><span class="w">   </span><span class="m">11.55182</span><span class="w">      </span><span class="m">2.3983021</span><span class="w"> </span><span class="m">0.70397545</span><span class="w">   </span><span class="m">2.952237</span><span class="w"> </span><span class="m">3.154811e-03</span><span class="w"> </span><span class="m">1.160401e-02</span><span class="w">
     </span><span class="m">45</span><span class="w"> </span><span class="m">1921.19192</span><span class="w">     </span><span class="m">-0.9062709</span><span class="w"> </span><span class="m">0.08444206</span><span class="w">  </span><span class="m">-6.942877</span><span class="w"> </span><span class="m">3.841942e-12</span><span class="w"> </span><span class="m">3.594718e-11</span><span class="w">
</span></code></pre></div></div>

<h3 id="visualizing-results">
<a class="anchor" href="#visualizing-results" aria-hidden="true"><span class="octicon octicon-link"></span></a>Visualizing results</h3>

<p>In addition to the MA plot explored previously, another useful plot providing a global view of the results is the volcano plot, which shows the fold changes relative to the adjusted p-values for all genes. First, create a column of logical values indicating if the gene is DE using the mutate() function, with p-adjusted value threshold less than 0-point-05. Then, use ggplot2 to plot the log2 foldchange values versus the -log10 adjusted p-value. The points for the genes should then be colored by whether they are significant using the threshold column.</p>

<p>The volcano plot helps us to get an idea of the range of fold changes needed to identify significance in our data.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Generate logical column </span><span class="w">
</span><span class="n">smoc2_res_all</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">smoc2_results</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="n">mutate</span><span class="p">(</span><span class="n">threshold</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">padj</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="m">0.05</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Create the volcano plot</span><span class="w">
</span><span class="n">ggplot</span><span class="p">(</span><span class="n">smoc2_res_all</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_point</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">log2FoldChange</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="o">-</span><span class="n">log10</span><span class="p">(</span><span class="n">padj</span><span class="p">),</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">threshold</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">xlab</span><span class="p">(</span><span class="s2">"log2 fold change"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">ylab</span><span class="p">(</span><span class="s2">"-log10 adjusted p-value"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">theme</span><span class="p">(</span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"none"</span><span class="p">,</span><span class="w"> 
        </span><span class="n">plot.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">rel</span><span class="p">(</span><span class="m">1.5</span><span class="p">),</span><span class="w"> </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">),</span><span class="w"> 
        </span><span class="n">axis.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">rel</span><span class="p">(</span><span class="m">1.25</span><span class="p">)))</span><span class="w">

     </span><span class="n">Warning</span><span class="o">:</span><span class="w"> </span><span class="n">Removed</span><span class="w"> </span><span class="m">25395</span><span class="w"> </span><span class="n">rows</span><span class="w"> </span><span class="n">containing</span><span class="w"> </span><span class="n">missing</span><span class="w"> </span><span class="n">values</span><span class="w"> </span><span class="p">(</span><span class="n">geom_point</span><span class="p">)</span><span class="n">.</span><span class="w">
</span></code></pre></div></div>

<p><img src="images/volcano.png" alt=""></p>

<h1 id="references">
<a class="anchor" href="#references" aria-hidden="true"><span class="octicon octicon-link"></span></a>References</h1>
<ol>
  <li>Bornstein P, Sage EH. Matricellular proteins: extracellular modulators of cell function. Curr Opin Cell Biol 2002; 14:608–616. doi: 10.1016/S0955-0674(02)00361-7</li>
  <li>Vannahme C, Gosling S, Paulsson M, Maurer P, Hartmann U. Characterization of SMOC-2, a modular extracellular calcium-binding protein. Biochem J 2003; 373:805–814. doi: 10.1042/bj20030532.</li>
</ol>

  </div><!-- from https://github.com/utterance/utterances -->
<script src="https://utteranc.es/client.js"
        repo="gdeol4/neongenes"
        issue-term="title"
        label="blogpost-comment"
        theme="github-light"
        crossorigin="anonymous"
        async>
</script><a class="u-url" href="/bioinformatics/r/2022/05/12/RNA-Seq-analysis-with-DESeq2.html" hidden></a>
</article>
      </div>
    </main><footer class="site-footer h-card">
  <data class="u-url" href="/"></data>

  <div class="wrapper">

    <div class="footer-col-wrapper">
      <div class="footer-col">
        <p class="feed-subscribe">
          <a href="/feed.xml">
            <svg class="svg-icon orange">
              <use xlink:href="/assets/minima-social-icons.svg#rss"></use>
            </svg><span>Subscribe</span>
          </a>
        </p>
      </div>
      <div class="footer-col">
        <p>Notes on machine learning, bioinformatics, and programming.</p>
      </div>
    </div>

    <div class="social-links"><ul class="social-media-list"><li><a rel="me" href="https://github.com/gdeol4" target="_blank" title="gdeol4"><svg class="svg-icon grey"><use xlink:href="/assets/minima-social-icons.svg#github"></use></svg></a></li><li><a rel="me" href="https://www.linkedin.com/in/gurkamal-deol-508675174" target="_blank" title="gurkamal-deol-508675174"><svg class="svg-icon grey"><use xlink:href="/assets/minima-social-icons.svg#linkedin"></use></svg></a></li></ul>
</div>

  </div>

</footer>
</body>

</html>
