<!DOCTYPE html>
<html lang="en"><head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="twitter:card" content="summary_large_image" /><!-- Begin Jekyll SEO tag v2.6.1 -->
<title>Introduction | Home</title>
<meta name="generator" content="Jekyll v4.1.1" />
<meta property="og:title" content="Introduction" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="Notes on machine learning, bioinformatics, and programming." />
<meta property="og:description" content="Notes on machine learning, bioinformatics, and programming." />
<link rel="canonical" href="https://www.gurkamal.com//2022/05/14/accessing-genbank.html" />
<meta property="og:url" content="https://www.gurkamal.com//2022/05/14/accessing-genbank.html" />
<meta property="og:site_name" content="Home" />
<meta property="og:type" content="article" />
<meta property="article:published_time" content="2022-05-14T00:00:00-05:00" />
<script type="application/ld+json">
{"url":"https://www.gurkamal.com//2022/05/14/accessing-genbank.html","@type":"BlogPosting","headline":"Introduction","dateModified":"2022-05-14T00:00:00-05:00","datePublished":"2022-05-14T00:00:00-05:00","mainEntityOfPage":{"@type":"WebPage","@id":"https://www.gurkamal.com//2022/05/14/accessing-genbank.html"},"description":"Notes on machine learning, bioinformatics, and programming.","@context":"https://schema.org"}</script>
<!-- End Jekyll SEO tag -->
<link rel="stylesheet" href="/assets/css/style.css"><link type="application/atom+xml" rel="alternate" href="https://www.gurkamal.com//feed.xml" title="Home" /><!-- the google_analytics_id gets auto inserted from the config file -->



<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-59WZ8TRRRR"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-59WZ8TRRRR');
</script>


<link rel="shortcut icon" type="image/x-icon" href="/images/favicon.ico"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/Primer/15.2.0/primer.css" integrity="sha512-xTz2ys4coGAOz8vuV1NcQBkgVmKhsSEtjbqyMJbBHRplFuvKIUo6xhLHpAyPt9mfR6twHJgn9OgVLuqOvjeBhg==" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.14.0/css/all.min.css" integrity="sha512-1PKOgIY59xJ8Co8+NE6FZ+LOAZKjy+KY8iq0G4B3CyeY6wYHN3yt9PW0XpSriVlkMXe40PTKnXrLnZ9+fkDaog==" crossorigin="anonymous" />
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.12.0/katex.min.css" integrity="sha512-h7nl+xz8wgDlNM4NqKEM4F1NkIRS17M9+uJwIGwuo8vGqIl4BhuCKdxjWEINm+xyrUjNCnK5dCrhM0sj+wTIXw==" crossorigin="anonymous" />
    <script src="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.12.0/katex.min.js" integrity="sha512-/CMIhXiDA3m2c9kzRyd97MTb3MC6OVnx4TElQ7fkkoRghwDf6gi41gaT1PwF270W6+J60uTmwgeRpNpJdRV6sg==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.12.0/contrib/auto-render.min.js" integrity="sha512-Do7uJAaHZm5OLrIv/yN4w0iG1dbu01kzdMNnFfu/mAqgUk6Nniv2JYHcwH+cNwjqgLcqcuBBk+JRvprLVI8azg==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha512-0doc9hKxR3PYwso42RD1p5ySZpzzuDiOwMrdCEh2WdJZCjcmFKc/wEnL+z8fBQrnHoiNWbo+3fiGkOYXBdQp4A==" crossorigin="anonymous"></script>
    <script>
    document.addEventListener("DOMContentLoaded", function() {
        renderMathInElement( document.body, {
        delimiters: [
            {left: "$$", right: "$$", display: true},
            {left: "[%", right: "%]", display: true},
            {left: "$", right: "$", display: false}
        ]}
        );
    });
    </script>


<script>
function wrap_img(fn) {
    if (document.attachEvent ? document.readyState === "complete" : document.readyState !== "loading") {
        var elements = document.querySelectorAll(".post img");
        Array.prototype.forEach.call(elements, function(el, i) {
            if (el.getAttribute("title") && (el.className != "emoji")) {
                const caption = document.createElement('figcaption');
                var node = document.createTextNode(el.getAttribute("title"));
                caption.appendChild(node);
                const wrapper = document.createElement('figure');
                wrapper.className = 'image';
                el.parentNode.insertBefore(wrapper, el);
                el.parentNode.removeChild(el);
                wrapper.appendChild(el);
                wrapper.appendChild(caption);
            }
        });
    } else { document.addEventListener('DOMContentLoaded', fn); }
}
window.onload = wrap_img;
</script>

<script>
    document.addEventListener("DOMContentLoaded", function(){
    // add link icon to anchor tags
    var elem = document.querySelectorAll(".anchor-link")
    elem.forEach(e => (e.innerHTML = '<i class="fas fa-link fa-xs"></i>'));
    });
</script>
</head>
<body><header class="site-header">

  <div class="wrapper"><a class="site-title" rel="author" href="/">Home</a><nav class="site-nav">
        <input type="checkbox" id="nav-trigger" class="nav-trigger" />
        <label for="nav-trigger">
          <span class="menu-icon">
            <svg viewBox="0 0 18 15" width="18px" height="15px">
              <path d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.032C17.335,0,18,0.665,18,1.484L18,1.484z M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.032C17.335,6.031,18,6.696,18,7.516L18,7.516z M18,13.516C18,14.335,17.335,15,16.516,15H1.484 C0.665,15,0,14.335,0,13.516l0,0c0-0.82,0.665-1.483,1.484-1.483h15.032C17.335,12.031,18,12.695,18,13.516L18,13.516z"/>
            </svg>
          </span>
        </label>

        <div class="trigger"><a class="page-link" href="/about/">About Me</a><a class="page-link" href="/search/">Search</a><a class="page-link" href="/categories/">Tags</a></div>
      </nav></div>
</header>
<main class="page-content" aria-label="Content">
      <div class="wrapper">
        <article class="post h-entry" itemscope itemtype="http://schema.org/BlogPosting">

  <header class="post-header">
    <h1 class="post-title p-name" itemprop="name headline">Introduction</h1><p class="post-meta post-meta-title"><time class="dt-published" datetime="2022-05-14T00:00:00-05:00" itemprop="datePublished">
        May 14, 2022
      </time>
       • <span class="read-time" title="Estimated read time">
    
    
      6 min read
    
</span></p>

    

    
      
        <div class="pb-5 d-flex flex-wrap flex-justify-end">
          <div class="px-2">

    <a href="https://github.com/gdeol4/neongenes/tree/master/_notebooks/2022-05-14-accessing-genbank.ipynb" role="button" target="_blank">
<img class="notebook-badge-image" src="/assets/badges/github.svg" alt="View On GitHub">
    </a>
</div>

          
          <div class="px-2">
    <a href="https://colab.research.google.com/github/gdeol4/neongenes/blob/master/_notebooks/2022-05-14-accessing-genbank.ipynb" target="_blank">
        <img class="notebook-badge-image" src="/assets/badges/colab.svg" alt="Open In Colab"/>
    </a>
</div>
        </div>
      </header>

  <div class="post-content e-content" itemprop="articleBody">
    <!--
#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: _notebooks/2022-05-14-accessing-genbank.ipynb
-->

<div class="container" id="notebook-container">
        
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>This tutorial uses the NCBI API to interface with Entrez. To get started, the neccesarry libraries need to be imported and an email needs to be provided (so NCBI can contact you about your query if needed).</p>
<p>The goal will be to find the chloroquine resistance transporter (CRT) gene in the parasite Plasmodium flaciparum in the nucleotide database.</p>

</div>
</div>
</div>
    
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span><span class="p">,</span> <span class="n">Medline</span><span class="p">,</span> <span class="n">SeqIO</span>
</pre></div>

    </div>
</div>
</div>

</div>
    

    
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">email_address</span> <span class="o">=</span> <span class="s2">&quot;gurkamal.dev@gmail.com&quot;</span> 
</pre></div>

    </div>
</div>
</div>

</div>
    

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Handles are mentioned quite frequently throughout this documentation, and are also fairly confusing (at least to me!). Basically, you can think of a handle as being a “wrapper” around text information.</p>
<p>Handles provide (at least) two benefits over plain text information:</p>
<p>They provide a standard way to deal with information stored in different ways. The text information can be in a file, or in a string stored in memory, or the output from a command line program, or at some remote website, but the handle provides a common way of dealing with information in all of these formats.
They allow text information to be read incrementally, instead of all at once. This is really important when you are dealing with huge text files which would use up all of your memory if you had to load them all.
Handles can deal with text information that is being read (e. ‍g. ‍reading from a file) or written (e. ‍g. ‍writing information to a file). In the case of a “read” handle, commonly used functions are read(), which reads the entire text information from the handle, and readline(), which reads information one line at a time. For “write” handles, the function write() is regularly used.</p>
<p>To see the available databases:</p>

</div>
</div>
</div>
    
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">einfo</span><span class="p">()</span>

<span class="c1">#Read and store the Entrez query record returned</span>
<span class="n">rec</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">rec</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>{&#39;DbList&#39;: [&#39;pubmed&#39;, &#39;protein&#39;, &#39;nuccore&#39;, &#39;ipg&#39;, &#39;nucleotide&#39;, &#39;structure&#39;, &#39;genome&#39;, &#39;annotinfo&#39;, &#39;assembly&#39;, &#39;bioproject&#39;, &#39;biosample&#39;, &#39;blastdbinfo&#39;, &#39;books&#39;, &#39;cdd&#39;, &#39;clinvar&#39;, &#39;gap&#39;, &#39;gapplus&#39;, &#39;grasp&#39;, &#39;dbvar&#39;, &#39;gene&#39;, &#39;gds&#39;, &#39;geoprofiles&#39;, &#39;homologene&#39;, &#39;medgen&#39;, &#39;mesh&#39;, &#39;ncbisearch&#39;, &#39;nlmcatalog&#39;, &#39;omim&#39;, &#39;orgtrack&#39;, &#39;pmc&#39;, &#39;popset&#39;, &#39;proteinclusters&#39;, &#39;pcassay&#39;, &#39;protfam&#39;, &#39;pccompound&#39;, &#39;pcsubstance&#39;, &#39;seqannot&#39;, &#39;snp&#39;, &#39;sra&#39;, &#39;taxonomy&#39;, &#39;biocollections&#39;, &#39;gtr&#39;]}
</pre>
</div>
</div>

</div>
</div>

</div>
    

    
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">esearch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;nucleotide&quot;</span><span class="p">,</span> <span class="n">term</span><span class="o">=</span><span class="s1">&#39;CRT[Gene Name] AND &quot;Plasmodium falciparum&quot;[Organism]&#39;</span><span class="p">)</span>

<span class="c1">#read the result that is returned</span>
<span class="n">rec_list</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    

    
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">rec_list</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>{&#39;Count&#39;: &#39;2022&#39;, &#39;RetMax&#39;: &#39;20&#39;, &#39;RetStart&#39;: &#39;0&#39;, &#39;IdList&#39;: [&#39;2196471109&#39;, &#39;2196471107&#39;, &#39;2196471105&#39;, &#39;2196471103&#39;, &#39;2196471101&#39;, &#39;2196471099&#39;, &#39;2196471097&#39;, &#39;2196471095&#39;, &#39;2196471093&#39;, &#39;2196471091&#39;, &#39;2196471089&#39;, &#39;2196471087&#39;, &#39;2196471085&#39;, &#39;2196471083&#39;, &#39;2196471081&#39;, &#39;2196471079&#39;, &#39;2196471077&#39;, &#39;2196471075&#39;, &#39;2196471073&#39;, &#39;2196471071&#39;], &#39;TranslationSet&#39;: [{&#39;From&#39;: &#39;&#34;Plasmodium falciparum&#34;[Organism]&#39;, &#39;To&#39;: &#39;&#34;Plasmodium falciparum&#34;[Organism]&#39;}], &#39;TranslationStack&#39;: [{&#39;Term&#39;: &#39;CRT[Gene Name]&#39;, &#39;Field&#39;: &#39;Gene Name&#39;, &#39;Count&#39;: &#39;4778&#39;, &#39;Explode&#39;: &#39;N&#39;}, {&#39;Term&#39;: &#39;&#34;Plasmodium falciparum&#34;[Organism]&#39;, &#39;Field&#39;: &#39;Organism&#39;, &#39;Count&#39;: &#39;258609&#39;, &#39;Explode&#39;: &#39;Y&#39;}, &#39;AND&#39;], &#39;QueryTranslation&#39;: &#39;CRT[Gene Name] AND &#34;Plasmodium falciparum&#34;[Organism]&#39;}</pre>
</div>

</div>

</div>
</div>

</div>
    

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>The standard search will limit the number of record references to 20, so if you have more, you may want to repeat the query with an increased maximum limit. In this case, we will actually override the default limit with retmax. The Entrez system provides quite a few sophisticated ways to retrieve large number of results (for more information, check the Biopython or NCBI Entrez documentation). Although you now have the IDs of all of the records, you still need to retrieve the records properly.</p>

</div>
</div>
</div>
    
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">if</span> <span class="n">rec_list</span><span class="p">[</span><span class="s1">&#39;RetMax&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">rec_list</span><span class="p">[</span><span class="s1">&#39;Count&#39;</span><span class="p">]:</span>
    <span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">esearch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;nucleotide&quot;</span><span class="p">,</span> <span class="n">term</span><span class="o">=</span><span class="s1">&#39;CRT[Gene Name] AND &quot;Plasmodium falciparum&quot;[Organism]&#39;</span><span class="p">,</span>
                            <span class="n">retmax</span><span class="o">=</span><span class="n">rec_list</span><span class="p">[</span><span class="s1">&#39;Count&#39;</span><span class="p">])</span>
    <span class="n">rec_list</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>be careful with this technique, because you will retrieve a large amount of complete records, and some of them will have fairly large sequences inside.</p>

</div>
</div>
</div>
    
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">id_list</span> <span class="o">=</span> <span class="n">rec_list</span><span class="p">[</span><span class="s1">&#39;IdList&#39;</span><span class="p">]</span>

<span class="n">handle_2</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s1">&#39;nucleotide&#39;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="n">id_list</span><span class="p">,</span> <span class="n">rettype</span><span class="o">=</span><span class="s1">&#39;gb&#39;</span><span class="p">,</span> <span class="n">retmax</span><span class="o">=</span><span class="n">rec_list</span><span class="p">[</span><span class="s1">&#39;Count&#39;</span><span class="p">])</span>
</pre></div>

    </div>
</div>
</div>

</div>
    

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>will retrieve a list of records in the GenBank format (which includessequences, plus a lot of interesting metadata).Let's read and parse the result</p>

</div>
</div>
</div>
    
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">recs</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">handle_2</span><span class="p">,</span> <span class="s1">&#39;gb&#39;</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

</div>
    

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Note that we have converted the result of <code>SeqIO.parse</code>, an iterator, to a list. TheSeqIO.parse advantage of doing this is that we can use the result as many times as we want(for example, iterate many times over), without repeating the query on the server.This saves time, bandwidth, and server usage if you plan to iterate many timesover.</p>
<p>The disadvantage is that it will allocate memory for all records. This will not work for very large datasets</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>The <code>rec</code>  variable now has our record of interest. The rec.description file will contain its human-readable description.</p>

</div>
</div>
</div>
    
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">for</span> <span class="n">rec</span> <span class="ow">in</span> <span class="n">recs</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">rec</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="s1">&#39;KM288867&#39;</span><span class="p">:</span>
        <span class="k">break</span>
<span class="nb">print</span><span class="p">(</span><span class="n">rec</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">rec</span><span class="o">.</span><span class="n">description</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>KM288867
Plasmodium falciparum clone PF3D7_0709000 chloroquine resistance transporter (CRT) gene, complete cds
</pre>
</div>
</div>

</div>
</div>

</div>
    

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Now, let's extract some sequence features, which contain information such as gene products and exon positions on the sequence:</p>

</div>
</div>
</div>
    
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">rec</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">feature</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s1">&#39;gene&#39;</span><span class="p">:</span>
        <span class="c1"># gene name will be in the feature.qualifiers dictionary.</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">])</span>
        <span class="c1"># print the start, end, and originating strand of the exon</span>
    <span class="k">elif</span> <span class="n">feature</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s1">&#39;exon&#39;</span><span class="p">:</span>
        <span class="n">loc</span> <span class="o">=</span> <span class="n">feature</span><span class="o">.</span><span class="n">location</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Exon&#39;</span><span class="p">,</span> <span class="n">loc</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">loc</span><span class="o">.</span><span class="n">end</span><span class="p">,</span> <span class="n">loc</span><span class="o">.</span><span class="n">strand</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;not processed:</span><span class="se">\n</span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">feature</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>not processed:
type: source
location: [0:10000](+)
qualifiers:
    Key: clone, Value: [&#39;PF3D7_0709000&#39;]
    Key: db_xref, Value: [&#39;taxon:5833&#39;]
    Key: mol_type, Value: [&#39;genomic DNA&#39;]
    Key: organism, Value: [&#39;Plasmodium falciparum&#39;]

[&#39;CRT&#39;]
not processed:
type: mRNA
location: join{[2751:3543](+), [3720:3989](+), [4168:4341](+), [4513:4646](+), [4799:4871](+), [4994:5070](+), [5166:5249](+), [5376:5427](+), [5564:5621](+), [5769:5862](+), [6055:6100](+), [6247:6302](+), [6471:7598](+)}
qualifiers:
    Key: gene, Value: [&#39;CRT&#39;]
    Key: product, Value: [&#39;chloroquine resistance transporter&#39;]

not processed:
type: 5&#39;UTR
location: [2751:3452](+)
qualifiers:
    Key: gene, Value: [&#39;CRT&#39;]

not processed:
type: primer_bind
location: [2935:2958](+)
qualifiers:

not processed:
type: primer_bind
location: [3094:3121](+)
qualifiers:

not processed:
type: CDS
location: join{[3452:3543](+), [3720:3989](+), [4168:4341](+), [4513:4646](+), [4799:4871](+), [4994:5070](+), [5166:5249](+), [5376:5427](+), [5564:5621](+), [5769:5862](+), [6055:6100](+), [6247:6302](+), [6471:6548](+)}
qualifiers:
    Key: codon_start, Value: [&#39;1&#39;]
    Key: gene, Value: [&#39;CRT&#39;]
    Key: product, Value: [&#39;chloroquine resistance transporter&#39;]
    Key: protein_id, Value: [&#39;AIW62921.1&#39;]
    Key: translation, Value: [&#39;MKFASKKNNQKNSSKNDERYRELDNLVQEGNGSRLGGGSCLGKCAHVFKLIFKEIKDNIFIYILSIIYLSVCVMNKIFAKRTLNKIGNYSFVTSETHNFICMIMFFIVYSLFGNKKGNSKERHRSFNLQFFAISMLDACSVILAFIGLTRTTGNIQSFVLQLSIPINMFFCFLILRYRYHLYNYLGAVIIVVTIALVEMKLSFETQEENSIIFNLVLISALIPVCFSNMTREIVFKKYKIDILRLNAMVSFFQLFTSCLILPVYTLPFLKQLHLPYNEIWTNIKNGFACLFLGRNTVVENCGLGMAKLCDDCDGAWKTFALFSFFNICDNLITSYIIDKFSTMTYTIVSCIQGPAIAIAYYFKFLAGDVVREPRLLDFVTLFGYLFGSIIYRVGNIILERKKMRNEENEDSEGELTNVDSIITQ&#39;]

Exon 3452 3543 1
Exon 3720 3989 1
Exon 4168 4341 1
not processed:
type: primer_bind
location: [4288:4323](-)
qualifiers:

Exon 4513 4646 1
Exon 4799 4871 1
Exon 4994 5070 1
Exon 5166 5249 1
Exon 5376 5427 1
Exon 5564 5621 1
Exon 5769 5862 1
Exon 6055 6100 1
Exon 6247 6302 1
Exon 6471 6548 1
not processed:
type: 3&#39;UTR
location: [6548:7598](+)
qualifiers:
    Key: gene, Value: [&#39;CRT&#39;]

not processed:
type: primer_bind
location: [7833:7856](-)
qualifiers:

</pre>
</div>
</div>

</div>
</div>

</div>
    

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>We will also print all the locations of exons. Exons, as withall features, have locations in this sequence: a start, an end, and the strand fromwhere they are read. While all the start and end positions for our exons areExactPosition, note that Biopython supports many other types of positions.One type of position is , which specifies that a location point isBeforePositionbefore a certain sequence position. Another type of position isBetweenPosition, which gives the interval for a certain location start/end. Thereare quite a few more position types; these are just some examples.Coordinates will be specified in such a way that you will be able to retrieve thesequence from a Python array with ranges easily, so generally, the start will beone before the value on the record, and the end will be equal.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>We will now look at the annotations on the record, which are mostly metadata that is not related to the sequence position. Note that some values are not strings; they can be numbers or even lists (forexample, the taxonomy annotation is a list)</p>

</div>
</div>
</div>
    
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">rec</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">%s</span><span class="s1">=</span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">value</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>molecule_type=DNA
topology=linear
data_file_division=INV
date=12-NOV-2014
accessions=[&#39;KM288867&#39;]
sequence_version=1
keywords=[&#39;&#39;]
source=Plasmodium falciparum (malaria parasite P. falciparum)
organism=Plasmodium falciparum
taxonomy=[&#39;Eukaryota&#39;, &#39;Sar&#39;, &#39;Alveolata&#39;, &#39;Apicomplexa&#39;, &#39;Aconoidasida&#39;, &#39;Haemosporida&#39;, &#39;Plasmodiidae&#39;, &#39;Plasmodium&#39;, &#39;Plasmodium (Laverania)&#39;]
references=[Reference(title=&#39;Versatile control of Plasmodium falciparum gene expression with an inducible protein-RNA interaction&#39;, ...), Reference(title=&#39;Direct Submission&#39;, ...)]
</pre>
</div>
</div>

</div>
</div>

</div>
    

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Last but not least, you can access the fundamental piece of information, the sequence</p>

</div>
</div>
</div>
    
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">rec</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>10000
</pre>
</div>
</div>

</div>
</div>

</div>
    

    
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">rec</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">100</span><span class="p">])</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>ATATGTAAAACCAAAATAAATTAAACAGAATTTATTTTTAAAAGATTTATTTGTAACAATATTACCATGATGATTTATTAAAGTAAAATCACCACCTATT
</pre>
</div>
</div>

</div>
</div>

</div>
    

</div>



  </div><a class="u-url" href="/2022/05/14/accessing-genbank.html" hidden></a>
</article>
      </div>
    </main><footer class="site-footer h-card">
  <data class="u-url" href="/"></data>

  <div class="wrapper">

    <div class="footer-col-wrapper">
      <div class="footer-col">
        <p class="feed-subscribe">
          <a href="/feed.xml">
            <svg class="svg-icon orange">
              <use xlink:href="/assets/minima-social-icons.svg#rss"></use>
            </svg><span>Subscribe</span>
          </a>
        </p>
      </div>
      <div class="footer-col">
        <p>Notes on machine learning, bioinformatics, and programming.</p>
      </div>
    </div>

    <div class="social-links"><ul class="social-media-list"><li><a rel="me" href="https://github.com/gdeol4" target="_blank" title="gdeol4"><svg class="svg-icon grey"><use xlink:href="/assets/minima-social-icons.svg#github"></use></svg></a></li><li><a rel="me" href="https://www.linkedin.com/in/gurkamal-deol-508675174" target="_blank" title="gurkamal-deol-508675174"><svg class="svg-icon grey"><use xlink:href="/assets/minima-social-icons.svg#linkedin"></use></svg></a></li></ul>
</div>

  </div>

</footer>
</body>

</html>
